File: agent-governance-golang/packages/agentmesh/sandbox_test.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package agentmesh

import (
	"testing"
)

// Compile-time interface satisfaction check.
var _ SandboxProvider = (*DockerSandboxProvider)(nil)

func TestDefaultSandboxConfig(t *testing.T) {
	cfg := DefaultSandboxConfig()

	if cfg.TimeoutSeconds != 60 {
		t.Errorf("TimeoutSeconds = %v, want 60", cfg.TimeoutSeconds)
	}
	if cfg.MemoryMB != 512 {
		t.Errorf("MemoryMB = %v, want 512", cfg.MemoryMB)
	}
	if cfg.CPULimit != 1.0 {
		t.Errorf("CPULimit = %v, want 1.0", cfg.CPULimit)
	}
	if cfg.NetworkEnabled {
		t.Error("NetworkEnabled = true, want false")
	}
	if !cfg.ReadOnlyFS {
		t.Error("ReadOnlyFS = false, want true")
	}
	if cfg.EnvVars == nil {
		t.Error("EnvVars is nil, want initialized map")
	}
	if len(cfg.EnvVars) != 0 {
		t.Errorf("EnvVars length = %d, want 0", len(cfg.EnvVars))
	}
}

func TestDockerSandboxProviderNew(t *testing.T) {
	p := NewDockerSandboxProvider("alpine:latest")

	if p.image != "alpine:latest" {
		t.Errorf("image = %q, want %q", p.image, "alpine:latest")
	}
	if p.containers == nil {
		t.Error("containers map is nil, want initialized map")
	}
	// available may be true or false depending on the host; just verify no panic.
	t.Logf("Docker available: %v", p.IsAvailable())
}

func TestSandboxProviderInterface(t *testing.T) {
	// This test verifies that DockerSandboxProvider satisfies SandboxProvider at compile time.
	// The var _ declaration above is the actual check; this test ensures it is exercised.
	var provider SandboxProvider = NewDockerSandboxProvider("alpine:latest")
	if provider == nil {
		t.Fatal("provider should not be nil")
	}
}

func TestCreateExecuteDestroy(t *testing.T) {
	p := NewDockerSandboxProvider("alpine:latest")
	if !p.IsAvailable() {
		t.Skip("Docker is not available, skipping integration test")
	}

	cfg := DefaultSandboxConfig()
	session, err := p.CreateSession("test-agent", cfg)
	if err != nil {
		t.Fatalf("CreateSession failed: %v", err)
	}
	if session.AgentID != "test-agent" {
		t.Errorf("AgentID = %q, want %q", session.AgentID, "test-agent")
	}
	if session.Status != SessionRunning {
		t.Errorf("Status = %q, want %q", session.Status, SessionRunning)
	}

	handle, err := p.ExecuteCode("test-agent", session.SessionID, "echo hello")
	if err != nil {
		t.Fatalf("ExecuteCode failed: %v", err)
	}
	if !handle.Result.Success {
		t.Errorf("Success = false, want true; stderr: %s", handle.Result.Stderr)
	}
	if handle.Result.Stdout != "hello\n" {
		t.Errorf("Stdout = %q, want %q", handle.Result.Stdout, "hello\n")
	}
	if handle.Status != ExecutionCompleted {
		t.Errorf("Status = %q, want %q", handle.Status, ExecutionCompleted)
	}

	err = p.DestroySession("test-agent", session.SessionID)
	if err != nil {
		t.Fatalf("DestroySession failed: %v", err)
	}
}

File: agent-governance-typescript/tests/sandbox.test.ts
// Copyright (c) Microsoft Corporation. Licensed under the MIT License.

import {
  defaultSandboxConfig,
  DockerSandboxProvider,
  SessionStatus,
  ExecutionStatus,
} from '../src/sandbox';

describe('defaultSandboxConfig', () => {
  it('returns expected default values', () => {
    const cfg = defaultSandboxConfig();
    expect(cfg.timeoutSeconds).toBe(60);
    expect(cfg.memoryMb).toBe(512);
    expect(cfg.cpuLimit).toBe(1.0);
    expect(cfg.networkEnabled).toBe(false);
    expect(cfg.readOnlyFs).toBe(true);
    expect(cfg.envVars).toEqual({});
  });

  it('returns a fresh object each call', () => {
    const a = defaultSandboxConfig();
    const b = defaultSandboxConfig();
    expect(a).not.toBe(b);
    a.envVars['FOO'] = 'bar';
    expect(b.envVars).toEqual({});
  });
});

describe('DockerSandboxProvider', () => {
  it('uses default image python:3.11-slim', () => {
    const provider = new DockerSandboxProvider();
    expect(provider).toBeDefined();
    // Verify it's an instance implementing SandboxProvider
    expect(typeof provider.createSession).toBe('function');
    expect(typeof provider.executeCode).toBe('function');
    expect(typeof provider.destroySession).toBe('function');
    expect(typeof provider.isAvailable).toBe('function');
  });

  it('accepts a custom image', () => {
    const provider = new DockerSandboxProvider('node:20-slim');
    expect(provider).toBeDefined();
  });

  it('isAvailable handles missing Docker gracefully', async () => {
    const provider = new DockerSandboxProvider();
    // Should not throw — returns boolean regardless of Docker presence
    const result = await provider.isAvailable();
    expect(typeof result).toBe('boolean');
  });
});

describe('SessionStatus enum', () => {
  it('has expected members', () => {
    expect(SessionStatus.Provisioning).toBe('provisioning');
    expect(SessionStatus.Ready).toBe('ready');
    expect(SessionStatus.Executing).toBe('executing');
    expect(SessionStatus.Destroying).toBe('destroying');
    expect(SessionStatus.Destroyed).toBe('destroyed');
    expect(SessionStatus.Failed).toBe('failed');
  });
});

describe('ExecutionStatus enum', () => {
  it('has expected members', () => {
    expect(ExecutionStatus.Pending).toBe('pending');
    expect(ExecutionStatus.Running).toBe('running');
    expect(ExecutionStatus.Completed).toBe('completed');
    expect(ExecutionStatus.Cancelled).toBe('cancelled');
    expect(ExecutionStatus.Failed).toBe('failed');
  });
});

// Full lifecycle test — skip when Docker is not available
describe('DockerSandboxProvider lifecycle', () => {
  let provider: DockerSandboxProvider;
  let dockerAvailable: boolean;

  beforeAll(async () => {
    provider = new DockerSandboxProvider();
    dockerAvailable = await provider.isAvailable();
  });

  it('creates, executes, and destroys a session', async () => {
    if (!dockerAvailable) {
      console.log('Skipping lifecycle test — Docker not available');
      return;
    }

    // Create session
    const session = await provider.createSession('testAgent');
    expect(session.agentId).toBe('testAgent');
    expect(session.sessionId).toBeTruthy();
    expect(session.status).toBe(SessionStatus.Ready);

    try {
      // Execute code
      const handle = await provider.executeCode(
        'testAgent',
        session.sessionId,
        'print("hello sandbox")',
      );
      expect(handle.executionId).toBeTruthy();
      expect(handle.agentId).toBe('testAgent');
      expect(handle.sessionId).toBe(session.sessionId);
      expect(handle.status).toBe(ExecutionStatus.Completed);
      expect(handle.result).toBeDefined();
      expect(handle.result!.success).toBe(true);
      expect(handle.result!.stdout.trim()).toBe('hello sandbox');
      expect(handle.result!.exitCode).toBe(0);
      expect(handle.result!.durationSeconds).toBeGreaterThan(0);
    } finally {
      // Always clean up
      await provider.destroySession('testAgent', session.sessionId);
    }
  }, 60_000);

  it('destroySession is idempotent', async () => {
    if (!dockerAvailable) {
      console.log('Skipping idempotent test — Docker not available');
      return;
    }

    // Destroying a non-existent session should not throw
    await expect(
      provider.destroySession('testAgent', 'nonexistent-id'),
    ).resolves.toBeUndefined();
  });

  it('executeCode rejects for unknown session', async () => {
    await expect(
      provider.executeCode('testAgent', 'nonexistent-id', 'print("hi")'),
    ).rejects.toThrow(/No active session/);
  });
});
